### ISSUE TITLE
Support for ElasticSearch or SOLR

### ISSUE THREAD

**Issue by travisbotello** [Sterc/SimpleSearch#12]
We are running a MODx installation with about 40,000 resources (mainly articles). As you can imagine a single search request using SimpleSearch without ElasticSearch or SOLR takes some time which doesn't satisfy the user experience. Since the SOLR and ElastiSearch drivers have been removed from SimpleSearch 2.x we are looking for assistance/help on how to efficiently power SimpleSearch with one of those search drivers...

Any help/suggestions is very appreciated.

**Comment by travisbotello** [comment]
I ended up cloning modxcms/simeplesearch repo and building a custom version 1.9.3 because in 1.9.2/2.x some very interesting commits were never included. Now SimpleSearch is running with ElasticSearch without any problems and the performance is also very impressive. I'm not sure if it was the right step to completely remove the drivers, snippets and plugins for ElasticSearch/Solr in 2.x branch because they seem to still work pretty well...I'd love to implement those into 2.x branch but I don't have the time to do this right now.

**Comment by sdrenth** [comment]
@travisbotello We removed the ElasticSearch and SOLR driver because we thought that these drivers would not work anymore and since we never use the ElasticSearch/SOLR drivers we simply could test if these where still working. We would be glad to bring those drivers back into a new release of SimpleSearch (including the indexer ofcourse).

Would you be willing to help us out by testing our PR if we bring these drivers back into SimpleSearch to see if the drivers are still working fine?

**Comment by travisbotello** [comment]
@sdrenth Absolutely...right now I am successfully running Elastic 2.4.6 with my custom SimpleSearch version 1.9.3. So I would be definitely interested to get it running with latest Elastic version. I already gave it a try with Elastic 6.5.1. Therefore I updated elastica PHP client to highest version which lead to some deprecated methods inside the current elastic search driver. I tried to update those methods but was running out of time and went back to 2.4.6. But I am sure that it won't be that tricky to migrate it to newest version...

**Comment by sdrenth** [comment]
@travisbotello That would be great! I've readded the plugin and ES/SOLR drivers in this branch:
https://github.com/Sterc/SimpleSearch/tree/es-solr-support-12

Would you like to test the ES/SOLR drivers? If so, you can fork the branch use Git Package Management (http://theboxer.github.io/Git-Package-Management/directory-structure/) for installing the Extra and testing it.

I did notice that when I change the `simplesearch.driver_class` to `SimpleSearchDriverElastic` that the plugin gives a 500 error:
`Uncaught Error: Class 'SimpleSearchDriverElastic_mysql' not found in`.

I couldn't find any mysql classes for the SimpleSearchDriverElastic, so you have to generate these yourself or something?

**Comment by travisbotello** [comment]
First thing I notice is that `Psr` lib is missing in your branch. That is a dependency of ES.https://github.com/modxcms/SimpleSearch/tree/master/core/components/simplesearch/model/simplesearch/driver/libs 

Also the namespace has changed from `sisea` to `simplesearch` and in the driver classes  the old namespace is still used. Maybe that is related to the error message that you are seeing. I don't think I generated that class myself. Actually it should just use the `simplesearchdriverbasic.class.php` in the mysql folder.

Thanks for setting up that branch. When I have the time I will check it and test it on our staging. Right now we are about to go live with 1.9.3 and ES. The speed for plus 40k resources it absolutely stunning. So that should give us a huge benefit to get it working with current SimpleSearch version and current ES version.

**Comment by sdrenth** [comment]
@travisbotello That would be great! I added the missing lib and also fixed the namespace issue.

I managed to fix the 500 error by setting the `simplesearch.driver_class_path` to no.

**Comment by travisbotello** [comment]
That is good news! Thanks! Wondering what the default for `simplesearch.driver_class_path` was because in my installation it is empty.

**Comment by sdrenth** [comment]
@travisbotello I just added the system settings for the drivers aswell. 

The `simplesearch.driver_class_path` path should be empty, it is an optional path you can set if you want to store your drivers outside of the simplesearch directory for example.

**Comment by travisbotello** [comment]
@sdrenth Sorry for late reply, but I was super busy the last couple of month. I just setup the https://github.com/Sterc/SimpleSearch/tree/es-solr-support-12 branch in my local env and first time using `GPM`. When adding SimpleSearch via external directory I am receiving the following error in GPM. Can I just skip it or is this something breaking?

![Screenshot 2019-08-07 15 01 35](https://user-images.githubusercontent.com/1157380/62625081-cf562000-b924-11e9-9955-b7891b7cf757.png)

**Comment by travisbotello** [comment]
Please review https://github.com/Sterc/SimpleSearch/pull/33

This PR should make SimpleSearch ready for usage with newest ElasticSearch 7.X

**Comment by travisbotello** [comment]
@sdrenth Any plans on when this is going to be merged?

**Comment by sdrenth** [comment]
@travisbotello I've just merged the PR. 

From now on, we want to make extras more open source, since we don't have the resources to maintain all extras. 

Therefore I'll invite you as a contributor. You are free to merge to release a new version and if you have a MODX.com account you are able to upload the new version on there as well. 

This way we want to stimulate the development of the extras.

**Comment by travisbotello** [comment]
@sdrenth Thanks a lot for your quick reply!!
